torch/csrc/autograd/anomaly_mode.cpp

```diff
@@ -0,0 +1,49 @@
+#include "torch/csrc/python_headers.h"
+#include "torch/csrc/autograd/anomaly_mode.h"
+#include "torch/csrc/utils/auto_gil.h"
+#include "torch/csrc/utils/python_strings.h"
+#include "torch/csrc/utils/object_ptr.h"
+#include "torch/csrc/Exceptions.h"
+
+#include <stdexcept>
+#include <iostream>
+
+namespace torch { namespace autograd {
+
+bool AnomalyMode::_enabled = 0;
+PyObject* AnomalyMode::stacktrace_key = THPUtils_packString("stacktrace_");
+
+void AnomalyMode::store_stack(PyObject* metadata) {
+  AutoGIL gil;
+  THPObjectPtr mod(PyImport_ImportModule("traceback"));
+  if (!mod) {
+    throw python_error();
+  }
+
+  THPObjectPtr list(PyObject_CallMethod(mod.get(), "format_stack", ""));
+
+  if (PyDict_SetItem(metadata, AnomalyMode::stacktrace_key, list.get())) {
+    throw python_error();
+  }
+}
+
+void AnomalyMode::print_stack(PyObject* metadata) {
+  AutoGIL gil;
+  if (!PyDict_Check(metadata)) {
+    throw std::runtime_error("Function metadata is not a python dictionary.");
+  }
+
+  if (!PyDict_Contains(metadata, AnomalyMode::stacktrace_key)) {
+    std::cout << "No forward pass information available." << std::endl;
+    std::cout << "Enable detect anomaly during forward pass for more informations." << std::endl;
+    return;
+  }
+
+  THPObjectPtr empty_string(PyUnicode_FromString(""));
+  THPObjectPtr msg(PyUnicode_Join(empty_string, PyDict_GetItem(metadata, AnomalyMode::stacktrace_key)));
+
+  std::cout << "Traceback of forward call that caused the error:" << std::endl;
+  std::cout << THPUtils_unpackString(msg) << std::endl;
```

Comment:
Use of cout here makes me a little nervous; this is just begging for weird output buffering issues with Python's stdout error buffering.

Reply:
I agree but I did that because the exception that we find can be many things: pure python, our custom python, pure cpp. In the two last cases, I can wrap them in a new custom error and add extra text to them with the new informations. But for the pure python one, I don't know how to do it or if it's even possible.
Even though this is not ideal, this is the simplest implementation I found.
If you know how to tinker with any possible native python error to change it's message to add a string to it, I can do the change.

Reply:
Or as Adam suggested, just make this a warning :)